src/database.ts

```diff
@@ -1,92 +1,290 @@
+  return Object.entries(set)
+    .sort(([a], [b]) => a.localeCompare(b))
+    .map(([k, v]) => ({ [k]: v }));
+}
+
+export function keyConfigOpts(sthis: SuperThis, name?: string, opts?: ConfigOpts): string {
+  return JSON.stringify(
+    toSortedArray({
+      name,
+      stores: toSortedArray(JSON.parse(JSON.stringify(toStoreURIRuntime(sthis, name, opts?.storeUrls)))),
+    }),
+  );
+}
+
+export interface DatabaseOpts {
   readonly name?: string;
-  readonly opts: ConfigOpts = {};
+  // readonly public?: boolean;
+  readonly meta?: DbMeta;
+  // readonly factoryUnreg?: () => void;
+  // readonly persistIndexes?: boolean;
+  // readonly autoCompact?: number;
+  readonly storeUrls: StoreURIRuntime;
+  readonly storeEnDe: StoreEnDeFile;
+  readonly keyBag: KeyBagRuntime;
+  // readonly threshold?: number;
+}
+
+export interface Database<DT extends DocTypes = NonNullable<unknown>> extends HasCRDT<DT> {
+  // readonly name: string;
+  readonly logger: Logger;
+  readonly sthis: SuperThis;
+  readonly id: string;
+
+  name(): string;
+
+  onClosed(fn: () => void): void;
+
+  close(): Promise<void>;
+  destroy(): Promise<void>;
+  ready(): Promise<void>;
+
+  get<T extends DocTypes>(id: string): Promise<DocWithId<T>>;
+  put<T extends DocTypes>(doc: DocSet<T>): Promise<DocResponse>;
+  del(id: string): Promise<DocResponse>;
+  changes<T extends DocTypes>(since?: ClockHead, opts?: ChangesOptions): Promise<ChangesResponse<T>>;
+  allDocs<T extends DocTypes>(opts?: AllDocsQueryOpts): Promise<AllDocsResponse<T>>;
+  allDocuments<T extends DocTypes>(): Promise<{
+    rows: {
+      key: string;
+      value: DocWithId<T>;
+    }[];
+    clock: ClockHead;
+  }>;
+  subscribe<T extends DocTypes>(listener: ListenerFn<T>, updates?: boolean): () => void;
+
+  query<K extends IndexKeyType, T extends DocTypes, R extends DocFragment = T>(
+    field: string | MapFn<T>,
+    opts?: QueryOpts<K>,
+  ): Promise<IndexRows<K, T, R>>;
+  compact(): Promise<void>;
+}
+
+export function isDatabase<T extends DocTypes = NonNullable<unknown>>(db: unknown): db is Database<T> {
+  return db instanceof DatabaseImpl || db instanceof DatabaseShell;
+}
+
+export function DatabaseFactory<T extends DocTypes = NonNullable<unknown>>(
+  name: string | undefined,
+  opts?: ConfigOpts,
+): Database<T> {
+  const sthis = ensureSuperThis(opts);
+  return new DatabaseShell<T>(
+    databases.get(keyConfigOpts(sthis, name, opts)).once((key) => {
+      const db = new DatabaseImpl<T>(sthis, {
+        name,
+        meta: opts?.meta,
+        keyBag: defaultKeyBagOpts(sthis, opts?.keyBag),
+        storeUrls: toStoreURIRuntime(sthis, name, opts?.storeUrls),
+        storeEnDe: {
+          encodeFile,
+          decodeFile,
+          ...opts?.storeEnDe,
+        },
+      });
+      db.onClosed(() => {
+        databases.unget(key);
+      });
+      return db;
+    }),
+  );
+}
+
+export class DatabaseShell<DT extends DocTypes = NonNullable<unknown>> implements Database<DT> {
```

Comment:
I wanna understand what motivates this layer

Reply:
close need to be refcounted 
if you open multiple times the same database like:
fireproof("test")
fireproof("test")
fireproof("test")
fireproof("test")
then close should do nothing until the last instance is closed